dom/src/main/scala/fs2/dom/MediaSource.scala

```diff
@@ -0,0 +1,303 @@
+package fs2.dom
+
+import org.scalajs.dom
+import scala.scalajs.js
+import cats.effect._
+import cats.syntax.all._
+
+import scala.scalajs.js.annotation.*
+
+object Url {
+
+  private def createObjectUrl[F[_]](
+      objectUrl: F[String]
+  )(implicit F: Async[F]): Resource[F, Url] =
+    Resource
+      .make(objectUrl)(url => F.delay(dom.URL.revokeObjectURL(url)))
+      .map(url => new dom.URL(url))
+
+  def createObjectUrl[F[_]](blob: dom.Blob)(implicit F: Async[F]): Resource[F, Url] =
+    createObjectUrl(F.delay(dom.URL.createObjectURL(blob)))
+
+  def createObjectUrl[F[_]](src: MediaSource[F])(implicit F: Async[F]): Resource[F, Url] =
+    createObjectUrl(F.delay(dom.URL.createObjectURL(src)))
+}
```

Comment:
I wonder if we should take the `Dom` constraint instead of `Async`.

Reply:
I wasn't sure what the `Dom` constraint was for? Is it meant to be a capability; and in which case, do we expect a wrapper trait for each of the different API's such as the web audio context or media extensions?

Reply:
That's a good question. So the problem with an `Async` or `Sync` constraint is that, from the user-side, if you see that constraint then it means that this code could be doing almost anything.

However, if you see a `Dom` constraint, then (in theory) it limits the scope of what that code is doing. I guess in practice it doesn't really matter ...

Reply:
Hmm. I think there is some value in it, but I think that we need to be clear what the semantics of what it should mean.

We can either:

1. Add the `Dom`/`WebAudio` on the _constructors_ of anything which will "touch the dom", "create an audio context". The objects then act as a witness for the rest of the code, so if you pass around an audio context afterwards it means that you're going to be using "WebAudio" probably if you have an `F`.
2. Add a specific capability on the actual methods so `AudioContext.resume` for example needs the capability.

Regardless, I'm not certain `createObjectUrl` would fit into either of these brackets. It doesn't touch the `Dom` at all and it essentially just creates a root to an object and passes back a weak reference (`String`) to be used in other calls?

Reply:
Sorry, actually I take that back. I check MDN and it says that `createObjectUrl` is available in web workers. But `window` is not available in web workers. So that wouldn't make sense.

Let's leave it for now, I'll think more about it.